File: day-01/solution.ts
import { readLines } from "../commons/io.ts";

function main() {
  const lines = readLines("day-01/input.txt");
  let answer = 0;
  lines.forEach((line) => {
    const newLine = stringNumReplacer(line);
    const sum = findLineSum(newLine);
    // console.log(line, " -> ", newLine, " -> ", sum)
    answer += sum;
  });

  console.log("answer:", answer);
}

function stringNumReplacer(line: string) {
  let newLine = line;
  newLine = newLine.replaceAll("one", "o1ne");
  newLine = newLine.replaceAll("two", "t2wo");
  newLine = newLine.replaceAll("three", "th3ree");
  newLine = newLine.replaceAll("four", "fo4ur");
  newLine = newLine.replaceAll("five", "fi5ve");
  newLine = newLine.replaceAll("six", "s6ix");
  newLine = newLine.replaceAll("seven", "se7ven");
  newLine = newLine.replaceAll("eight", "ei8ght");
  newLine = newLine.replaceAll("nine", "ni9ne");
  return newLine;
}

function findLineSum(line: string) {
  const chars = line.split("");
  let leftPtr = 0,
    rightPtr = chars.length - 1;
  let leftNum = 0,
    rightNum = 0;
  for (;;) {
    const char = chars[leftPtr];
    // check if char is num
    const num = parseInt(char);
    if (Number.isNaN(num)) {
      leftPtr++;
      continue;
    } else {
      leftNum = num;
      break;
    }
  }
  for (;;) {
    const char = chars[rightPtr];
    // check if char is num
    const num = parseInt(char);
    if (Number.isNaN(num)) {
      rightPtr--;
      continue;
    } else {
      rightNum = num;
      break;
    }
  }
  return Number(`${leftNum}${rightNum}`);
}

main();

File: day-01/solution.go
package main

import (
	cio "commons"
	"fmt"
	"strconv"
	"strings"
)

func main() {
	lines := cio.ReadLines("day-01/input.txt")
	answer := 0
	for _, line := range lines {
		newLine := stringNumReplacer(line)
		sum := findLineSum(newLine)
		answer += sum
	}
	fmt.Println("answer:", answer)
}

func stringNumReplacer(line string) string {
	newLine := line
	newLine = strings.ReplaceAll(newLine, "one", "o1ne")
	newLine = strings.ReplaceAll(newLine, "two", "t2wo")
	newLine = strings.ReplaceAll(newLine, "three", "th3ree")
	newLine = strings.ReplaceAll(newLine, "four", "fo4ur")
	newLine = strings.ReplaceAll(newLine, "five", "fi5ve")
	newLine = strings.ReplaceAll(newLine, "six", "s6ix")
	newLine = strings.ReplaceAll(newLine, "seven", "se7ven")
	newLine = strings.ReplaceAll(newLine, "eight", "ei8ght")
	newLine = strings.ReplaceAll(newLine, "nine", "ni9ne")
	return newLine
}

func findLineSum(line string) int {
	chars := strings.Split(line, "")
	leftPtr := 0
	rightPtr := len(chars) - 1
	leftNum := 0
	rightNum := 0
	for {
		char := chars[leftPtr]
		num, err := strconv.Atoi(char)
		if err != nil {
			leftPtr++
			continue
		} else {
			leftNum = num
			break
		}
	}
	for {
		char := chars[rightPtr]
		num, err := strconv.Atoi(char)
		if err != nil {
			rightPtr--
			continue
		} else {
			rightNum = num
			break
		}
	}
	return leftNum*10 + rightNum
}
